### ISSUE TITLE
Rule proposal: resolver not called in Promise executor

### ISSUE THREAD

**Issue by lcrocker** [eslint/eslint#9691]
The executor function passed to new Promise() must at some point call the resolver function that is passed into it. For example:

return new Promise((resolve, reject) => {
    return 5;
});

Should trigger the error.

This happens to me occasionally when I convert synchronous code into promise form; this will also catch misspellings of the resolver.

**Comment by j-f1** [comment]
I made [something like this](https://github.com/j-f1/eslint-plugin-desktop/blob/master/lib/rules/promise-must-complete.js) in a plugin that I haven’t released to `npm` yet. ([read the docs for it 📖](https://github.com/j-f1/eslint-plugin-desktop/blob/master/docs/rules/promise-must-complete.md)) I’d be happy to turn this into a PR for ESLint core.

PS: If anyone at @eslint is interested in any of [the other rules](https://github.com/j-f1/eslint-plugin-desktop#supported-rules), I’d be very glad to get those to a wider audience, too. Just @j-f1 me and I’ll open PRs.

**Comment by not-an-aardvark** [comment]
Would this be covered by `no-unused-vars` already?

**Comment by j-f1** [comment]
@not-an-aardvark 

> Would this be covered by `no-unused-vars` already?

Not in this case:

```js
new Promise((resolve, reject) => {
  if (myCondition()) {
    resolve(42)
  } else if (myOtherCondition()) {
    reject(NaN)
  } else {
    // forgot to call `resolve()` or `reject()` here
    doSomething()
  }
})
```

**Comment by platinumazure** [comment]
Ah, looks like this would need to use code path analysis to ensure the resolver (or rejection function?) is called in every code path.

I think I could :+1: this.

**Comment by lcrocker** [comment]
I guess no-unused-vars would catch my simplified example. Here's the actual code that caught me:

	static removeLocalFirmwareFile(name) {
		let	dd = cordova.file.dataDirectory;
		return new Promise((resolve, reject) => {
			window.resolveLocalFileSystemURL(dd, system => {
				system.getDirectory('Firmware', {
					create: true,
					exclusive: false
				}, directory => {
					directory.getFile(name, {
						create: false
					}, entry => {
						entry.remove(() => {
							// resolve(true);
							return true;
						}, reject);
					}, reject);
				}, reject);
			}, reject);
		});
	}

This did not trigger "no-unused-vars" despite no "resolve" appearing anywhere.

**Comment by j-f1** [comment]
> This did not trigger "no-unused-vars" despite no "resolve" appearing anywhere.

I think this is by design (you may not want that parameter in another callback), but it makes `no-unused-vars` kinda useless here.

**Comment by mysticatea** [comment]
You can use `{args: "all"}` option to catch the `resolve` parameter.

[Demo](https://mysticatea.github.io/vue-eslint-demo/#eMKcbVLDi27DgjAQw7wVKxcCCsKJw5QjL8KpasOFwolDVcOaU8OTwoNlL8OIwpVjR8OrTcKAwqLDvHttSMKABg7CscKyM8Kzw689RsOCSsKIJsORw4wJVCUtcsOMTTYCwqfClSFmw6zCuDLClQM5wq45wroJw7sCRMKLCTtyw5x6wpNrw518wrNRFnzDggfDu8OSIsKxTWUEKWsYQmFrWFnDgcO1UmHCscOjCEvCpSE2wrzCgCE7wp7DnEgDw6UkJcKbM2FRw5rCmsKnG8KvSSUnw77CqhAEWTxMw49SBMKqw5AwAzvDtsKGwrZQDsOiGMOBWV1Dw6LCk8O9eMOxwpDDjRddw6TCnHbDikjCu0tbTVvCiMKGw7XDgREUwp/Dr8KrWMOKwoTCucKTw7XDjy/CpzPCmG7CgS5Fw4TCg8KuwolBcivDjUkgcMKCCSPCrCDCuSVgL3TDpVTDrcK5DcOXDsKuXMKTMMOZBcOuwqXDtsOTw6jCmMKQw78ywrFew5Jrw5pew6hTcDDDtCDCsMKvKMOgw6l5MXE8fMKgw4gpw4tYO8KxOHQ0wpzDtgXDrR4Cw5nDo8Kaw4sewqbCvcKCw67DoQfDoB3DlMKcf8Kaw7DDjMKyw7ZEwqMkw4JKwoPCiybDhybCicO8wop9U2vDtcOrwq/DuMKpMz8OZThqV3IBXg5Sw7lxwrZgOMO4ZyPCl2oPw7IlHH/DswfDr8OUHhI=)

I'm using the rule with `{args: "all", argsIgnorePattern: "/^_./"}` to ensure unused parameters explicitly.

**Comment by platinumazure** [comment]
It might still be good to ensure all code paths call, or pass as callback functions, one of the resolve/reject functions. Sort of like consistent-return maybe? no-unused-vars does not (and should not) check invocations in code paths.

**Comment by mysticatea** [comment]
I agreed. However, I thought that it's difficult (sorry, I couldn't find the issue).

I guess that the rule checks whether either `resolve` or `reject` is used in all path.

```js
static removeLocalFirmwareFile(name) {
	let	dd = cordova.file.dataDirectory;
	return new Promise((resolve, reject) => {
		window.resolveLocalFileSystemURL(dd, system => { ... }, reject);
	});
}
```

But in this case, `reject` is used in all path.

**Comment by platinumazure** [comment]
Hmm... Maybe a new rule would check the following:

* `resolve` and `reject` are each used at least once (or resolve only, if reject isn't used in the Promise constructor); and
* All code paths execute exactly one of `resolve` or `reject`, exactly once (or, both are passed as callbacks to a function exactly once, or...?).

**Comment by not-an-aardvark** [comment]
It seems like the main use case for `new Promise` is creating a promise from another asynchronous task:

```js
new Promise((resolve, reject) => {
  fs.readFile(filename, (err, text) => {
    if (err) {
      reject(err);
    } else {
      resolve(text);
    }
  });
});
```

But code path analysis wouldn't catch this case, because it wouldn't be able to tell whether the callback to `fs.readFile` will ever be called (and the variable is used as a closure).

It seems like code path analysis would only be able to catch cases where `resolve` and `reject` are used synchronously, e.g.

```js
const x = new Promise((resolve, reject) => {
  if (foo) {
    resolve(5);
  } else {
    reject('error');
  }
});
```

But in cases like that, it's probably better to not use the `Promise` constructor anyway:

```js
const x = foo ? Promise.resolve(5) : Promise.reject('error');
```

**Comment by mysticatea** [comment]
Because there are some cases that it overs the border of functions, it's interesting work. For example:

```js
function get(name) {
    return new Promise((resolve, reject) => {
        if (cache.has(name)) {
            resolve(cache.get(id))
            return
        }
        fs.readFile(name, "utf8", (err, content) => {
            if (err) {
                reject(err)
                return
            }
            cache.set(name, content)
            resolve(content)
        })
    })
}
```

----

There are valid cases even if neither `resolve` and `reject` in a path. Though it may be cases that `/*eslint-disable*/` is proper.

```js
function getAllInParallel(names) {
    return new Promise((resolve, reject) => {
        let results = new Map()

        for (const name of names) {
            if (cache.has(name)) {
                results.set(name, cache.get(id))
                continue
            }
            fs.readFile(name, "utf8", (err, content) => {
                if (err) {
                    reject(err)
                    return
                }

                results.set(name, content)
                if (results.size >= names.length) {
                    resolve(content)
                }
            })
        }
    })
}
```

```js
function readHead(name, size) {
    return new Promise((resolve, reject) => {
        const stream = fs.createReadStream(name)
        let result = ""

        stream.setEncoding("utf8")
        stream.on("data", (chunk) => {
            result += chunk
            if (result.length >= size) {
                stream.close()
                resolve(result.slice(0, size))
            }
        })
        stream.on("end", () => {
            resolve(result)
        })
        stream.on("error", reject)
    })
}
```

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
